File: api/stripe_webhook.py
import os
import stripe
from fastapi import APIRouter, Request, HTTPException
from starlette.responses import Response

router = APIRouter()
stripe.api_key = os.getenv("STRIPE_SECRET_KEY")
endpoint_secret = os.getenv("STRIPE_WEBHOOK_SECRET")

@router.post("/webhook")
async def stripe_webhook(request: Request):
    payload = await request.body()
    sig_header = request.headers.get("stripe-signature")

    try:
        event = stripe.Webhook.construct_event(
            payload, sig_header, endpoint_secret
        )
    except stripe.error.SignatureVerificationError:
        raise HTTPException(status_code=400, detail="Invalid signature")

    # 🔁 Handle subscription-related events here
    if event['type'] == 'checkout.session.completed':
        session = event['data']['object']
        print("✅ Payment received:", session["id"])
        # TODO: Activate license or log user

    return Response(status_code=200)

File: routes/stripe-webhook.js
const express = require('express');
const router = express.Router();
const Stripe = require('stripe');
const { createLicense } = require('../services/license');
require('dotenv').config();

const stripe = Stripe(process.env.STRIPE_SECRET_KEY);

router.post('/', express.raw({ type: 'application/json' }), (req, res) => {
  const sig = req.headers['stripe-signature'];
  let event;

  try {
    event = stripe.webhooks.constructEvent(
      req.body,
      sig,
      process.env.STRIPE_WEBHOOK_SECRET
    );
  } catch (err) {
    console.error('❌ Invalid webhook signature:', err.message);
    return res.status(400).send(`Webhook Error: ${err.message}`);
  }

  // 🎯 Handle completed checkout
  if (event.type === 'checkout.session.completed') {
    const session = event.data.object;
    const email = session.customer_details.email;
    const plan = session.metadata?.plan || 'unknown';

    console.log(`✅ Payment received from ${email} for ${plan}`);
    createLicense(email, plan);
  }

  res.json({ received: true });
});

module.exports = router;
